runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectMetrics.java

```diff
@@ -0,0 +1,298 @@
+    protected DistributionResult extract(DistributionData data) {
+      return data.extractResult();
+    }
+  }
+
+  /** The current values of counters in memory. */
+  private MetricsMap<MetricKey, DirectMetric<Long, Long>> counters =
+      new MetricsMap<>(new MetricsMap.Factory<MetricKey, DirectMetric<Long, Long>>() {
+        @Override
+        public DirectMetric<Long, Long> createInstance(MetricKey unusedKey) {
+          return new DirectCounter();
+        }
+      });
+  private MetricsMap<MetricKey, DirectMetric<DistributionData, DistributionResult>> distributions =
+      new MetricsMap<>(
+          new MetricsMap.Factory<MetricKey, DirectMetric<DistributionData, DistributionResult>>() {
+        @Override
+        public DirectMetric<DistributionData, DistributionResult> createInstance(
+            MetricKey unusedKey) {
+          return new DirectDistribution();
+        }
+      });
+
+  @AutoValue
+  abstract static class DirectMetricQueryResults implements MetricQueryResults {
+    public static MetricQueryResults create(
+        Iterable<MetricResult<Long>> counters,
+        Iterable<MetricResult<DistributionResult>> distributions) {
+      return new AutoValue_DirectMetrics_DirectMetricQueryResults(counters, distributions);
+    }
+  }
+
+  @AutoValue
+  abstract static class DirectMetricResult<T> implements MetricResult<T> {
+    public static <T> MetricResult<T> create(MetricName name, String scope,
+        T committed, T attempted) {
+      return new AutoValue_DirectMetrics_DirectMetricResult<T>(
+          name, scope, committed, attempted);
+    }
+  }
+
+  @Override
+  public MetricQueryResults queryMetrics(MetricsFilter filter) {
+    ImmutableList.Builder<MetricResult<Long>> counterResults = ImmutableList.builder();
+    for (Entry<MetricKey, DirectMetric<Long, Long>> counter : counters.entries()) {
+      maybeExtractResult(filter, counterResults, counter);
+    }
+    ImmutableList.Builder<MetricResult<DistributionResult>> distributionResults =
+        ImmutableList.builder();
+    for (Entry<MetricKey, DirectMetric<DistributionData, DistributionResult>> distribution
+        : distributions.entries()) {
+      maybeExtractResult(filter, distributionResults, distribution);
+    }
+
+    return DirectMetricQueryResults.create(counterResults.build(), distributionResults.build());
+  }
+
+  private <ResultT> void maybeExtractResult(
+      MetricsFilter filter,
+      ImmutableList.Builder<MetricResult<ResultT>> resultsBuilder,
+      Map.Entry<MetricKey, ? extends DirectMetric<?, ResultT>> entry) {
+    if (matches(filter, entry.getKey())) {
+      resultsBuilder.add(DirectMetricResult.create(
+          entry.getKey().metricName(),
+          entry.getKey().stepName(),
+          entry.getValue().extractCommitted(),
+          entry.getValue().extractLatestAttempted()));
+    }
+  }
+
+  private boolean matches(MetricsFilter filter, MetricKey key) {
+    return matchesName(key.metricName(), filter.names())
```

Comment:
This feels like the inverted responsibility for what I would expect `MetricsFilter` to have (especially when compared to the `Filter#by()` `PTransform`.)

MetricsFilter provides the names and steps to allow a runner to do better filtering without executing the user code, right? A reasoning comment to why it's just a value object rather than something with real behavior might be valuable here.